rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,221 @@
+
+Unlike targets, permissions typically originate from a fixed set. While the inclusion of wildcards, as seen in the `ALL` permission, can still be beneficial, doing it at Teleport-level may overlap with database-level functionality, potentially diminishing utility.
+
+### Permission Lifecycle
+
+The feature is based on the automated database user creation feature, simplifying the analysis of permission lifetime.
+
+We define "permission synchronization" as the process of applying a predetermined set of permissions to the user. These permissions are calculated based on their current roles in conjunction with the existing database schema.
+
+1. Initially, when the user does not exist, they possess no permissions except for default ones.
+2. Upon connection, we create a user in a manner specific to the database. This involves applying a fixed set of roles to the user, as specified in `spec.allow.db_roles`.
+3. Permission synchronization occurs after the user is created or reactivated.
+4. Additional permission synchronizations can be triggered as necessary, such as on new connections, detected schema changes (which may be detected using database-specific methods), or based on a scheduled timer.
+5. For consistency, we ensure that new connections receive an equivalent set of permissions. Notably, permissions for other protocols or unsupported versions do not affect what is applied to the database. Changes in the database schema can lead to previously "identical" permission sets diverging, for example, when a new table is added. The permissions are equivalent if they are stable under schema changes.
+
+6. Once the last connection for the user is terminated, they are either removed or deactivated.
+7. When a user is removed, all permissions are automatically revoked. In cases where full removal is not possible (example: a user is the owner of database objects in Postgres), a process is in place to ensure that all permissions are removed from the user, regardless of their source. This approach acts as a precaution against potential lingering permissions and simplifies the management of permissions.
+
+### TAG Integration
+
+Fundamentally, TAG operates on the basis of graph nodes interconnected by edges. To align our permissions with TAG, we'll represent each potential target as an individual node within the graph.
+
+Each target node will have its distinct set of relevant permissions, contingent on its type. For instance, tables may be associated with "SELECT" permissions, while procedures might involve "EXECUTE" permissions. These permissions will be represented as edges in the graph, each tagged with the appropriate permission type.
+
+The representation of allowed but unapplied permissions remains unclear within TAG. Expanding wildcards in permissions into individual nodes and edges is essential since TAG's representation is flat and explicit, lacking further interpretation.
```

Comment:
> If you move your permissions to the action, you don't need to replicate the permissions per database and the number of permutations will reduce significantly. 

Makes sense; I think we can have `(protocol type, permission type)` tuples, like `(Postgres, SELECT)`. This does materially reduce the number of nodes in the graph.

> You can also choose to represent the data inside the Database node instead of having external nodes so everything can be concise but I don't dislike to represent them as nodes.

This is an interesting option. Having a single Database node, instead of a number of nodes for each object in the database, would certainly be easier from the performance point of view. I think we would have to push the detailed permissions into the Action itself: `(Postgres, SELECT, "public.sales")` or `(Postgres, INSERT, "topsecret.*")`. This would need joining with the database schema to obtain an actual list of affected objects, but it would not be represented inside the graph itself.

I think this is a nice fallback scenario in case the "fully unfolded" option is unfeasible due to performance considerations. It would also work well with the policy-based permissions (row-level security), which are difficult to express in a graph.